File: js/menstruation.js
const prompt = require("prompt-sync")();

function monthCheck(month, months){

	if(month in months) return true
	return false

}


function dayCheck(day, month){

	if(isNaN(Number(day))) return false;

	day = Number(day);

	const array31 = {"january" : 0, "march" : 2, "may" : 4, "july" : 6, "august" : 7, "october" : 9, "december" : 11};

	const array30 = {"april" : 3,  "june" : 5, "september" : 8, "november" : 10};

	const array28 = {"february" : 1};

	if(month in array31){
		if(day > 0 && day < 32) return true;
		return false;
	}
	else
	if(month in array30){
		if(day > 0 && day < 31) return true;
		return false;
	}
	else
	if(month in array28){
		if(day > 0 && day < 29) return true;
		return false;
	}
	else return false;

}



function cycleCheck(cycle){

	if(isNaN(Number(cycle))) return false;
	return true;

}



function getOvulation(month, day, duration, cycle){

	const periodStart = new Date(2025, month, day);
	periodStart.setDate(periodStart.getDate() + (cycle/2));
	
	return periodStart;

}



function getFertilityWindow(month, day, duration, cycle){

	const periodStart = new Date(2025, month, day);
	const start = new Date(periodStart.setDate(periodStart.getDate() + ((cycle/2) - 5)));
	const end = new Date (periodStart.setDate(periodStart.getDate() + ((cycle/2) - 8)));
	
	return [start, end];

}



function getNextPeriod(month, day, duration, cycle){

	const periodStart = new Date(2025, month, day);
	periodStart.setDate(periodStart.getDate() + cycle);
	
	return periodStart;

}






const months = {"january" : 0, "february" : 1, "march" : 2, "april" : 3, "may" : 4, "june" : 5, "july" : 6, "august" : 7, "september" : 8, "october" : 9, "november" : 10, "december" : 11};



while(true){

	let pMonth = prompt("what month did your last period occur? ");
	pMonth = pMonth.toLowerCase();

	if(!monthCheck(pMonth, months)){
		console.log("invalid input");
		continue;
	}

	let pDay = prompt("what day did it start?(in digits): ");

	if(!dayCheck(pDay, pMonth)){
		console.log("invalid input");
		continue;
	}

	let pDuration = prompt("On average, How long does it last(in days): ");
	
	if(!cycleCheck(pDuration)){
		console.log("invalid input");
		continue;
	}

	let cycle = prompt("On average, how long is your menstrual cycle(in days): ");

	if(!cycleCheck(cycle)){
		console.log("invalid input");
		continue;
	}

	let monthIndex = months[pMonth]
	pDay = Number(pDay);
	pDuration = Number(pDuration);
	cycle = Number(cycle)	;


	let ovulationDay = getOvulation(monthIndex, pDay, pDuration, cycle);
	const [windowStart, windowEnd] = getFertilityWindow(monthIndex, pDay, pDuration, cycle);
	let nextPeriodStart = getNextPeriod(monthIndex, pDay, pDuration, cycle);


	console.log(`your ovulation day is on ${ovulationDay.toString()}`);
	console.log(`your fertile window is between ${windowStart.toString()} and ${windowEnd.toString()}`);
	console.log(`your next period starts on ${nextPeriodStart.toString()}`);
	break



}

File: menstruation/py/menstruation.py
import datetime
import calendar

def month_check(month, months):

	if month not in months: return False
	return True



def day_check(day, month):
	
	if not day.isdecimal():
		return False

	day = int(day)

	list_31 = ["January", "March", "May",  "July", "August", "October", "December"]
	list_30 = ["April", "June", "September", "November"]
	list_28 = ["February"]

	if month in list_31:
		if day in range(1, 32): return True
		return False
	elif month in list_30:
		if day in range(1, 31): return True
		return False
	elif month in list_28:
		if day in range(1, 29): return True
		return False
	else: return False




def cycle_check(cycle):

	if not cycle.isdecimal(): return False
	return True
	
	


def information(months):
	print()
	#calendar.prcal(2025)
	print()

	p_month = str(input("what month did your last period occur? ")).lower().capitalize()

	if not month_check(p_month, months): return "invalid input"

	print()
	calendar.prmonth(2025, months[p_month])
	print()

	p_day = str(input("what day did it start?(in digits): "))
	if not day_check(p_day, p_month): return "invalid input"

	p_duration = str(input("On average, How long does it last(in days): "))
	if not cycle_check(p_duration): return "invalid input"

	cycle = str(input("On average, how long is your menstrual cycle(in days): "))
	if not cycle_check(cycle): return "invalid input"


	p_day = int(p_day)
	p_duration = int(p_duration)
	cycle = int(cycle)	

	return (months[p_month], p_day, p_duration, cycle)




def get_ovulation(dates):

	month, day, duration, cycle = dates

	period_start = datetime.date(2025, month, day)
	fol_phase = datetime.timedelta(cycle/2)
	ovulation_day = period_start + fol_phase
	
	return ovulation_day.strftime("%B %d")




def get_fertility_window(dates):
	
	month, day, duration, cycle = dates

	period_start = datetime.date(2025, month, day)
	fol_phase = datetime.timedelta(cycle/2)
	ovulation_day = period_start + fol_phase

	window_start = ovulation_day - datetime.timedelta(5)
	window_end = ovulation_day + datetime.timedelta(1)

	return (window_start.strftime("%B %d"), window_end.strftime("%B %d"))




def get_next_period(dates):
	
	month, day, duration, cycle = dates

	period_start = datetime.date(2025, month, day)
	menstrual_cycle = datetime.timedelta(cycle)
	next_period_start = period_start + menstrual_cycle
	next_period_end = next_period_start + datetime.timedelta(duration)

	return (next_period_start.strftime("%B %d"), next_period_end.strftime("%B %d"))




	






